runtime/gc_structs/ClassArrayClassSlotIterator.cpp

```diff
@@ -42,6 +42,16 @@ GC_ClassArrayClassSlotIterator::nextSlot()
 	J9Class *classPtr = NULL;
 	while (_state != classArrayClassSlotIterator_state_done) {
 		switch (_state) {
+#if defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES)
+		case classArrayClassSlotIterator_state_nullRestrictedArrayClass:
+			classPtr = _iterateClazz->nullRestrictedArrayClass;
+			if (!_isArrayClass) {
+				_state = classArrayClassSlotIterator_state_done;
+			} else {
+				_state += 1;
+			}
+			break;
+#endif /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```

Comment:
Why you put this case first? From GC point of view does not matter how `nullRestrictedArrayClass` and `arrayClass` are relate, it just another field.
`if (!_isArrayClass)` should be located in the first case only (Iterator requested to walk non-array class.
So, I think you can remove `_isArrayClass` check, just keep state increment and move this case lower together with `classArrayClassSlotIterator_state_companionArray` case.

Reply:
Okay. So then I think the nullRestrictedArrayClass case isn't needed at all. The nullRestrictedArrayClass field will only contain an object for a ram class that is a base array. The same object would be accessible as the companionArray field of an array ram class.

Reply:
It is possible that a nullRestrictedArrayClass field exists when an arrayClass does not exist. Without the nullRestrictedArrayClass case this would be missed.

Reply:
Okay. The nullRestrictedArrayClass case will not be needed then since _iterateClazz is expected to be an array